### ISSUE TITLE
Edit script ad-hoc before execution

### ISSUE THREAD

**Issue by gergelyk** [casey/just#701]
How about an option of editing recipes ad-hoc, right before execution. E.g. having my `justfile`:

```
build:
  g++ main.cpp

run: build
  ./a.out
```

and invoking:
```sh
just --edit run
```
I would first get my default editor (e.g. from $EDITOR variable) open at some temporary file (e.g. /tmp/just.cmd) of following content:

```
# build
g++ main.cpp
# run
./a.out
```

Here I could do some fine tuning, e.g. add -O2 flag to `g++ main.cpp`. After saving the file and exiting from the editor, just would execute final commands from /tmp/just.cmd.

This idea would need some improvements for sake of mixed-language scripts though. E.g. variables of consecutive commands would fall into common name space.

**Comment by casey** [comment]
I like the idea, although it might be hard to implement well. I think a good first step, and maybe a useful feature on its own, would be to add the ability to "render" a justfile to a script that does the same thing as invoking Just with some set of args.

This might be used to "export" commands to run in environments where Just isn't present, or to allow editing commands before they are run. If such a feature were implemented, it would then be relatively easy to add the ability to edit a rendered script before running.

**Comment by gergelyk** [comment]
Exporting is a nice idea too, but introduces the same challenges as editing ad-hoc. Mixed-language script would need to look somehow like:

```sh
# first_command
cat <<EOF | /usr/bin/env python3
print('Hello from Python!')
EOF

# second_command
cat <<EOF | /usr/bin/env node
console.log('Greetings from JavaScript!')
EOF
```

And looks probably a bit over complicated. These features look easy to implement only for invocations where only singe recipe is called, I think.

**Comment by casey** [comment]
That's a good point, I didn't think about multi-language recipes.

**Comment by gergelyk** [comment]
I mean script of this kind:

```
python:
    #!/usr/bin/env python3
    print('Hello from python!')

js:
    #!/usr/bin/env node
    console.log('Greetings from JavaScript!')
```

Invoked like:

```sh
just --edit python js
```

Would result in something like:

```
#!/usr/bin/env python3
print('Hello from python!')

#!/usr/bin/env node
console.log('Greetings from JavaScript!')
```

which is not that practical. But... maybe some special type of delimiter line could separate recipes in edited/exported file. `just` could cut the file in partitions and consider each partition as a separate file to execute. E.g.:

```
#!/usr/bin/env python3
print('Hello from python!')
---
#!/usr/bin/env node
console.log('Greetings from JavaScript!')
```

**Comment by casey** [comment]
Hmm, thinking about it, I think the easiest way to do this would actually be to just create a temporary copy of the justfile somewhere, let the user edit it, and then run that.

**Comment by gergelyk** [comment]
I'm not sure. Probably this wouldn't be handy to navigate in a large justfile of a complex project. To be honest, I don't have any further ideas. Maybe this feature doesn't really fit to the entire design. Let's forget about it unless somebody comes up with a good suggestion.

**Comment by casey** [comment]
I agree that it's not ideal to navigate a large justfile if you just want to run one command. The advantage though is that a lot of the complexity goes away, since it can just be re-parsed with the existing parser.

In any, let's leave this open. I appreciate the suggestion!

**Comment by vmiheer** [comment]
On other hand, is there way to keep the generated shebang script for later inspection? I see a file created in `/tmp/justJOLujt/maestro` but it gets deleted when the just execution stops. Maybe printing the path at the beginning so that one can put `sleep 100` to inspect the content would be useful.

**Comment by casey** [comment]
I think this is probably too difficult to support well, relative to how niche it is.
